File: apps/api/src/app/api/v1/endpoints/tags.py
from __future__ import annotations

from sqlalchemy.exc import IntegrityError
from fastapi import APIRouter, Depends, HTTPException, Query, Response

from app.api.deps import get_tag_service
from app.api.security import require_internal_secret
from app.schemas.tag import TagCreate, TagRead, TagUpdate
from app.services.tag_service import TagInUseError, TagService, TagValidationError

router = APIRouter(dependencies=[Depends(require_internal_secret)])


@router.get(
    "",
    response_model=list[TagRead],
    summary="List tags",
    description=(
        "List or search tags for writer autosuggest and admin filtering. "
        "Requires x-internal-api-secret authentication header."
    ),
    responses={
        200: {"description": "Tags returned"},
        401: {"description": "Missing or invalid internal API secret"},
    },
)
def list_tags(
    query: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    service: TagService = Depends(get_tag_service),
) -> list[TagRead]:
    return service.list_tags(query=query, limit=limit, offset=offset)


@router.post(
    "",
    response_model=TagRead,
    summary="Create tag",
    description="Create a tag. Requires x-internal-api-secret authentication header.",
    responses={
        200: {"description": "Tag created"},
        400: {"description": "Tag payload is invalid"},
        401: {"description": "Missing or invalid internal API secret"},
        409: {"description": "Tag slug already exists"},
    },
)
def create_tag(
    payload: TagCreate,
    service: TagService = Depends(get_tag_service),
) -> TagRead:
    try:
        return service.create_tag(payload)
    except TagValidationError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except IntegrityError as exc:
        raise HTTPException(status_code=409, detail="tag slug already exists") from exc


@router.patch(
    "/{slug}",
    response_model=TagRead,
    summary="Update tag",
    description="Update a tag by slug. Requires x-internal-api-secret authentication header.",
    responses={
        200: {"description": "Tag updated"},
        400: {"description": "Tag payload is invalid"},
        401: {"description": "Missing or invalid internal API secret"},
        404: {"description": "Tag not found"},
        409: {"description": "Tag slug already exists"},
    },
)
def update_tag(
    slug: str,
    payload: TagUpdate,
    service: TagService = Depends(get_tag_service),
) -> TagRead:
    if payload.slug is None and payload.label is None:
        raise HTTPException(status_code=400, detail="at least one field is required")
    try:
        updated = service.update_tag(slug, payload)
    except TagValidationError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except IntegrityError as exc:
        raise HTTPException(status_code=409, detail="tag slug already exists") from exc
    if updated is None:
        raise HTTPException(status_code=404, detail="tag not found")
    return updated


@router.delete(
    "/{slug}",
    status_code=204,
    summary="Delete tag",
    description="Delete a tag by slug. Requires x-internal-api-secret authentication header.",
    responses={
        204: {"description": "Tag deleted"},
        401: {"description": "Missing or invalid internal API secret"},
        404: {"description": "Tag not found"},
        409: {"description": "Tag is linked to one or more posts"},
    },
)
def delete_tag(
    slug: str,
    force: bool = Query(default=False),
    service: TagService = Depends(get_tag_service),
) -> Response:
    try:
        deleted = service.delete_tag(slug, force=force)
    except TagInUseError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    if not deleted:
        raise HTTPException(status_code=404, detail="tag not found")
    return Response(status_code=204)

File: apps/api/src/routes/tags.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::Json,
};
use axum_extra::extract::Query;
use serde::Deserialize;
use utoipa::IntoParams;

use crate::{
    AppState,
    auth::RequireInternalSecret,
    error::{AppError, ErrorDetail},
    list_params::validate_limit_offset,
    posts::TagRead,
    tags::{TagCreate, TagUpdate, create_tag, delete_tag, list_tags, update_tag},
};

#[derive(Debug, Deserialize, IntoParams, Default)]
#[into_params(parameter_in = Query)]
pub struct ListTagsQuery {
    /// Free-text token matched against slug or label (case-insensitive).
    query: Option<String>,
    /// Page size (1..=200, default 50).
    limit: Option<i64>,
    /// Items skipped before this page (>= 0, default 0).
    offset: Option<i64>,
}

#[utoipa::path(
    get,
    path = "/tags",
    tag = "tags",
    operation_id = "list_tags",
    summary = "List tags",
    description = "List or search tags for writer autosuggest and admin filtering. Requires `x-internal-api-secret`.",
    params(ListTagsQuery),
    responses(
        (status = 200, description = "Tags returned", body = Vec<TagRead>),
        (status = 400, description = "Invalid query parameter", body = ErrorDetail),
        (status = 401, description = "Missing or invalid internal API secret", body = ErrorDetail),
        (status = 500, description = "Internal error", body = ErrorDetail),
    ),
    security(("internal_api_secret" = [])),
)]
pub async fn list_tags_handler(
    _: RequireInternalSecret,
    State(state): State<AppState>,
    Query(params): Query<ListTagsQuery>,
) -> Result<Json<Vec<TagRead>>, AppError> {
    let (limit, offset) = validate_limit_offset(params.limit, params.offset, 50, 200)?;
    let tags = list_tags(&state.pool, params.query.as_deref(), limit, offset).await?;
    Ok(Json(tags))
}

#[utoipa::path(
    post,
    path = "/tags",
    tag = "tags",
    operation_id = "create_tag",
    summary = "Create tag",
    description = "Create a new tag. The slug is normalized server-side.",
    request_body = TagCreate,
    responses(
        (status = 200, description = "Tag created", body = TagRead),
        (status = 400, description = "Tag payload is invalid", body = ErrorDetail),
        (status = 401, description = "Missing or invalid internal API secret", body = ErrorDetail),
        (status = 409, description = "Tag slug already exists", body = ErrorDetail),
        (status = 500, description = "Internal error", body = ErrorDetail),
    ),
    security(("internal_api_secret" = [])),
)]
pub async fn create_tag_handler(
    _: RequireInternalSecret,
    State(state): State<AppState>,
    Json(payload): Json<TagCreate>,
) -> Result<Json<TagRead>, AppError> {
    let tag = create_tag(&state.pool, payload).await?;
    Ok(Json(tag))
}

#[utoipa::path(
    patch,
    path = "/tags/{slug}",
    tag = "tags",
    operation_id = "update_tag",
    summary = "Update tag",
    description = "Update a tag's slug and/or label. At least one field is required.",
    params(
        ("slug" = String, Path, description = "Current tag slug"),
    ),
    request_body = TagUpdate,
    responses(
        (status = 200, description = "Tag updated", body = TagRead),
        (status = 400, description = "Tag payload is invalid", body = ErrorDetail),
        (status = 401, description = "Missing or invalid internal API secret", body = ErrorDetail),
        (status = 404, description = "Tag not found", body = ErrorDetail),
        (status = 409, description = "Tag slug already exists", body = ErrorDetail),
        (status = 500, description = "Internal error", body = ErrorDetail),
    ),
    security(("internal_api_secret" = [])),
)]
pub async fn update_tag_handler(
    _: RequireInternalSecret,
    State(state): State<AppState>,
    Path(slug): Path<String>,
    Json(payload): Json<TagUpdate>,
) -> Result<Json<TagRead>, AppError> {
    let tag = update_tag(&state.pool, &slug, payload)
        .await?
        .ok_or(AppError::NotFound("tag not found"))?;
    Ok(Json(tag))
}

#[derive(Debug, Deserialize, IntoParams, Default)]
#[into_params(parameter_in = Query)]
pub struct DeleteTagQuery {
    /// When `true`, removes the tag's post links before deleting; otherwise
    /// the request fails with 409 if any links exist.
    #[serde(default)]
    force: bool,
}

#[utoipa::path(
    delete,
    path = "/tags/{slug}",
    tag = "tags",
    operation_id = "delete_tag",
    summary = "Delete tag",
    description = "Delete a tag by slug. Use `?force=true` to detach existing post links first.",
    params(
        ("slug" = String, Path, description = "Tag slug to delete"),
        DeleteTagQuery,
    ),
    responses(
        (status = 204, description = "Tag deleted"),
        (status = 401, description = "Missing or invalid internal API secret", body = ErrorDetail),
        (status = 404, description = "Tag not found", body = ErrorDetail),
        (status = 409, description = "Tag is linked to one or more posts", body = ErrorDetail),
        (status = 500, description = "Internal error", body = ErrorDetail),
    ),
    security(("internal_api_secret" = [])),
)]
pub async fn delete_tag_handler(
    _: RequireInternalSecret,
    State(state): State<AppState>,
    Path(slug): Path<String>,
    Query(params): Query<DeleteTagQuery>,
) -> Result<StatusCode, AppError> {
    let deleted = delete_tag(&state.pool, &slug, params.force).await?;
    if !deleted {
        return Err(AppError::NotFound("tag not found"));
    }
    Ok(StatusCode::NO_CONTENT)
}
